### ISSUE TITLE
Circle build failures

### ISSUE THREAD

**Issue by kronosapiens** [JoinColony/colonyNetwork#791]
Within the last week (or so), our Circle builds (specifically `lint-and-unit-test`) have been failing repeatedly, with the following message in the Circle UI:

```
There was an issue while running this container and it was rerun. The most recent run is shown.
```

Whatever the issue, it is causing Circle to re-run the build, resulting in an endless loop of failing builds. Unfortunately, when Circle re-runs the build, the logging from the previous build is discarded, making the issue difficult to understand. The [help center article](https://support.circleci.com/hc/en-us/articles/360007792373-There-was-an-issue-while-running-this-container-and-it-was-rerun-The-most-recent-run-is-shown-) says that the issue is with Circle, and to open a support ticket (which I did).

Alex and I speculated that the issue might have been with a `ganache-cli` upgrade, but looking at `git blame`, we see that `ganache-cli` was last upgraded two months ago.

So that's that!

**Comment by area** [comment]
`yarn run test:contracts:upgrade:2to3` has stopped working, but #779 which removes that also doesn't build and shows this error. So I'm not sure what to conclude about that...

**Comment by kronosapiens** [comment]
Here is the support ticket:

```
##- Please type your reply above this line -##
Thank you for contacting CircleCI Support! Your request (#66503) has been received, and we'll be in touch soon.

To add additional comments, reply to this email.

We aim to reply promptly to all support requests. Please see below for Response Time SLAs for your support plan, and note that the priority of the issue will dictate the actual timeframe of the response. If you need a guaranteed response time, you may be interested in our Premium Support plans.

- Platinum: 1-2 hours
- Gold: 1-8 hours
- Silver: 2-24 hours
- Standard: No SLA guarantee (we typically do our best to respond within 3 business days)

-------------

Hi! I have been running into the "There was an issue while running this container and it was rerun. The most recent run is shown." issue for at least a week now, seemingly on every build in my project. The help center desk suggested submitting a support ticket, so here I am.
--
```

**Comment by kronosapiens** [comment]
Following on to Alex's comment, `yarn run test:contracts:upgrade:2to3` is broken, even on `develop`, with the error:

```
Error: while migrating Migrations: Returned error: Returned error: project ID does not have access to archive state
    at /Users/kronosapiens/code/colony/colonyNetwork/node_modules/truffle/build/webpack:/packages/deployer/src/deployment.js:364:1
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

**Comment by kronosapiens** [comment]
The [build](https://circleci.com/gh/JoinColony/colonyNetwork/10865) for #790 failed with the following (timeout error). For whatever reason this build was not re-run?

```
  50 passing (53m)
  3 pending
  5 failing

  1) Contract: Reputation Mining - happy paths
       when executing intended behaviours
         should cope if someone's existing reputation would go negative, setting it to zero instead:
     Error: timeout
      at Timeout._onTimeout (node_modules/ethers/utils/web.js:54:20)

  2) Contract: Reputation Mining - happy paths
       "after each" hook: after test for "should cope if someone's existing reputation would go negative, setting it to zero instead":
     Error: 
Could not connect to your Ethereum client with the following parameters:
    - host       > localhost
    - port       > 8545
    - network_id > 1580276322665
Please check that your Ethereum client:
    - is running
    - is accepting RPC connections (i.e., "--rpc" option is used in geth)
    - is accessible over the network
    - is properly configured in your Truffle configuration file (truffle-config.js)

      at ProviderError.ExtendableError (node_modules/truffle/build/webpack:/packages/error/index.js:10:1)
      at new ProviderError (node_modules/truffle/build/webpack:/packages/provider/error.js:12:1)
      at /home/circleci/colonyNetwork/node_modules/truffle/build/webpack:/packages/provider/wrapper.js:75:1
      at /home/circleci/colonyNetwork/node_modules/truffle/build/webpack:/packages/provider/wrapper.js:107:1
      at XMLHttpRequest.request.onreadystatechange (node_modules/truffle/build/webpack:/node_modules/web3-providers-http/src/index.js:96:1)
      at XMLHttpRequestEventTarget.dispatchEvent (node_modules/truffle/build/webpack:/node_modules/xhr2-cookies/dist/xml-http-request-event-target.js:34:1)
      at XMLHttpRequest._setReadyState (node_modules/truffle/build/webpack:/node_modules/xhr2-cookies/dist/xml-http-request.js:208:1)
      at XMLHttpRequest._onHttpRequestError (node_modules/truffle/build/webpack:/node_modules/xhr2-cookies/dist/xml-http-request.js:349:1)
      at ClientRequest.<anonymous> (node_modules/truffle/build/webpack:/node_modules/xhr2-cookies/dist/xml-http-request.js:252:47)
      at Socket.socketOnEnd (_http_client.js:425:9)
      at endReadableNT (_stream_readable.js:1094:12)
      at process._tickCallback (internal/process/next_tick.js:63:19)

  3) Contract: Reputation mining - root hash submissions
       "before all" hook: prepare suite:
     Error: Timeout of 120000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/circleci/colonyNetwork/test/reputation-system/root-hash-submissions.js)
  

  4) Contract: Reputation Mining - types of disagreement
       "before all" hook: prepare suite:
     Error: Timeout of 120000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/circleci/colonyNetwork/test/reputation-system/types-of-disagreement.js)
  

  5) Contract: Reputation miner client auto enabled functionality
       "before all" hook: prepare suite:
     Error: Timeout of 120000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/circleci/colonyNetwork/test/reputation-system/reputation-mining-client/client-auto-functionality.js)
```

**Comment by kronosapiens** [comment]
More notes:

- Re-building the current `develop` [fails](https://circleci.com/gh/JoinColony/colonyNetwork/10898) at `test:contracts:upgrade:2to3` due to (I believe) an Infura change in fetching archive data (see [here](https://infura.io/docs/ethereum/add-ons/archiveData)). Specifically, it does *not* seem to be timing out as the other tests are.

**Comment by area** [comment]
Okay, so this was a good one. Those two tests specifically failed because of the previous test, in classic async fashion. That test, unlike the others, did some checks after [`await miningCycleComplete`](https://github.com/JoinColony/colonyNetwork/blob/develop/test/reputation-system/reputation-mining-client/client-auto-functionality.js#L211). These caused it to hang around for about a minute, which resulted in `reputationMiningClient` possibly submitting a hash. 

This sometimes caused a knock-on effect in the next test, because of the way that block events and polling work in ethers. If blocks occurred after we had called `reputationMiningClient.close`, then when we brought up the mining client again, it fired events for all the blocks at once, starting with the earliest. The earliest block was the only one that ran `doBlockChecks` as for the other block events fired, `doBlockChecks` was already running. That meant that the mining client never saw that it was able to submit a hash (because that first block was too early) and because `testrpc only mines blocks if there is a transaction, the tests timed out.

We've fixed this by adding a check at the end of `doBlockChecks` (in `endDoBlockChecks`) that will run `doBlockChecks` again if one or more blocks were seen during the time `doBlockChecks` was run. We did not run it for every block in the interim, only the most recent one, as there is nothing in the checks that needs every block processed, or that cares about a historical state.
